const-oid/oiddbgen/src/root.rs

```diff
@@ -1,80 +1,38 @@
-use crate::{kind::Kind, node::Node, spec::Spec};
+use crate::{node::Node, spec::Spec};
 
-use std::collections::{BTreeMap, HashMap};
+use std::{collections::BTreeMap, ops::Deref};
 
 use proc_macro2::{Ident, Span, TokenStream};
 use quote::quote;
 
-#[derive(Clone, Debug)]
-pub struct Root {
-    tree: BTreeMap<u8, Kind>,
-    docs: HashMap<u8, (Ident, TokenStream)>,
-}
-
-impl Default for Root {
-    fn default() -> Self {
-        let mut docs = HashMap::new();
-
-        for (c, name, desc) in Self::DOCS {
-            let name = Ident::new(name, Span::call_site());
-            let desc = quote! { #![doc = #desc] };
-            docs.insert(*c, (name, desc));
-        }
-
-        let tree = BTreeMap::new();
-        Self { tree, docs }
-    }
-}
+#[derive(Clone, Debug, Default)]
+pub struct Root(BTreeMap<Ident, Spec>);
 
 impl Root {
-    const DOCS: &'static [(u8, &'static str, &'static str)] = &[
-        // From IANA
-        (b'A', "attr", "Attribute Type"),
-        (b'C', "dit", "DIT Content Rule"),
-        (b'E', "url", "LDAP URL Extension"),
-        (b'F', "family", "Family"),
-        (b'M', "matching", "Matching Rule"),
-        (b'N', "name", "Name Form"),
-        (b'O', "obj", "Object Class"),
-        (b'R', "admin", "Administrative Role"),
-        // Custom Additions
-        (b'X', "ext", "X.509 Certificate Extensions"),
-    ];
-
-    pub fn parse_line(&mut self, line: &str) {
-        let (name, next) = line.split_at(line.find(',').unwrap());
-        let (kind, next) = next[1..].split_at(next[1..].find(',').unwrap());
-        let (obji, spec) = next[1..].split_at(next[1..].find(',').unwrap());
-
-        let arc: Option<usize> = obji.find('.').and_then(|i| obji.split_at(i).0.parse().ok());
-        if arc.is_some() && spec.trim().starts_with(",[RFC") {
-            let name = name.trim().to_string();
-            let kind = kind.trim().as_bytes()[0];
-            let obji = obji.trim().to_string();
-            let spec = Ident::new(
-                &spec[2..][..spec.len() - 3].to_ascii_lowercase(),
-                Span::call_site(),
-            );
-
-            if self.docs.contains_key(&kind) {
-                self.tree
-                    .entry(kind)
-                    .or_insert_with(Kind::default)
-                    .entry(spec)
-                    .or_insert_with(Spec::default)
-                    .push(Node::new(obji, name));
-            }
-        }
+    pub fn add(
+        &mut self,
+        spec: impl Deref<Target = str>,
+        name: impl Deref<Target = str>,
+        obid: impl Deref<Target = str>,
```

Comment:
Implicit deref coercion should make the `impl Deref`s unnecessary:

```suggestion
        spec: &str,
        name: &str,
        obid: &str,
```

Reply:
It forces `main.rs` to take a reference. But I don't care either way.

Reply:
Fixed.

Reply:
IMO it's pretty much the main reason the `Deref` trait exists: so "smart pointer" types that impl `Deref` can be treated as regular references